src/Microsoft.DotNet.Build.Tasks/GatherFoldersToRestore.cs

```diff
@@ -0,0 +1,73 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Text;
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class GatherDirectoriesToRestore : Task
+    {
+        [Required]
+        public string[] RootDirectories { get; set; }
+
+        [Output]
+        public string[] DirectoriesToRestore { get; set; }
+
+        public int DirectoryListLength = 7000;
+
+        public override bool Execute()
+        {
+            HashSet<string> directoriesToRestore = new HashSet<string>();
+
+            foreach (string rootDirectory in RootDirectories)
+            {
+                AddDirectoriesToRestore(rootDirectory, rootDirectory, directoriesToRestore);
+            }
+
+            List<string> resultBuilder = new List<string>();
+            StringBuilder sb = new StringBuilder(DirectoryListLength);
+
+            foreach (string directoryPath in directoriesToRestore)
+            {
+                sb.Append('"');
+                sb.Append(directoryPath);
+                sb.Append("\" ");
+
+                if (sb.Length >= DirectoryListLength)
+                {
+                    resultBuilder.Add(sb.ToString());
+                    sb.Clear();
+                }
+            }
+
+            if (sb.Length > 0)
+            {
+                resultBuilder.Add(sb.ToString());
+            }
+
+            DirectoriesToRestore = resultBuilder.ToArray();
+
+            return true;
+        }
+
+        public void AddDirectoriesToRestore(string directoryToExplore, string directoryToAdd, HashSet<string> directoriesToRestore)
+        {
+            bool hasProjectJson = File.Exists(Path.Combine(directoryToExplore, "project.json"));
+
+            if (hasProjectJson)
+            {
+                directoriesToRestore.Add(directoryToAdd);
+            }
+
+            foreach (string childDirectory in Directory.GetDirectories(directoryToExplore))
+            {
+                AddDirectoriesToRestore(childDirectory, hasProjectJson ? childDirectory : directoryToAdd, directoriesToRestore);
```

Comment:
Yeah the idea was to find root directories.  If there's a project.json, it means that whatever parent directory got us here needed to be restored.  If there isn't one, restoring the parent directory is fine.

For example:

```
Foo
Foo\Bar\project.json
Foo\Bar\Baz\project.json
Foo\Baz\project.json
```

In this case, we call the function with DirectoryToAdd as "Foo".  There's no project.json, so we recur to both Bar and Baz but keep DirectoryToAdd as "Foo"

When we Explore Foo\Bar, we see a project.json, so we add "Foo" (since DNU's recursive walk on Foo will find the project.json in the same directory) and when we recur in to the child directories we see "directoryToAdd" to the new child path (in this case Foo\Bar\Baz).

So we'd end up with `Foo` and `Foo\Bar\Baz` and the directories to restore.